File: app/src/main/java/com/jds/fitnessjunkiess/getfitapp/Data/ViewModels/ExerciseViewModel.java
package com.jds.fitnessjunkiess.getfitapp.Data.ViewModels;

import android.app.Application;
import android.arch.lifecycle.AndroidViewModel;
import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MediatorLiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.Transformations;

import com.jds.fitnessjunkiess.getfitapp.Data.DataModels.Exercise;
import com.jds.fitnessjunkiess.getfitapp.Data.Repositories.ExercisesRepository;
import com.jds.fitnessjunkiess.getfitapp.Pojo.ExercisesFilter;

import java.util.List;

public class ExerciseViewModel extends AndroidViewModel {
  ExercisesRepository repository;
  MediatorLiveData<List<Exercise>> data = new MediatorLiveData<>();

  private MutableLiveData<ExercisesFilter> filterMutableLiveData;

  public ExerciseViewModel(Application application) {
    super(application);
    this.repository = new ExercisesRepository(application);
    this.filterMutableLiveData = new MutableLiveData<>();

    ExercisesFilter filter = new ExercisesFilter("", "");
    this.filterMutableLiveData.setValue(filter);

    LiveData<List<Exercise>> source = Transformations.switchMap(
        this.filterMutableLiveData, value -> repository.filterSelect(value.muscleGroup, value.type)
    );

    this.data.addSource(source, val -> this.data.setValue(val));
  }

  public void setFilters(String muscleGroups, String type) {

  }

  public LiveData<List<Exercise>> select() {
    return data;
  }

  public void insert(Exercise exercise) {
    this.repository.insert(exercise);
  }

  public void setFilterMutableLiveData(ExercisesFilter exercisesFilter) {
    this.filterMutableLiveData.setValue(exercisesFilter);
  }
}

File: app/src/main/java/com/jds/fitnessjunkiess/getfitapp/data/viewModels/ExerciseViewModel.kt
package com.jds.fitnessjunkiess.getfitapp.data.viewModels

import android.app.Application
import android.arch.lifecycle.AndroidViewModel
import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MediatorLiveData
import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.Transformations

import com.jds.fitnessjunkiess.getfitapp.data.dataModels.Exercise
import com.jds.fitnessjunkiess.getfitapp.data.repositories.ExercisesRepository
import com.jds.fitnessjunkiess.getfitapp.data.pojo.ExercisesFilters

class ExerciseViewModel(application: Application) : AndroidViewModel(application) {

    private var repository: ExercisesRepository = ExercisesRepository(application)
    internal var data = MediatorLiveData<List<Exercise>>()

    private val filterMutableLiveData: MutableLiveData<ExercisesFilters> = MutableLiveData()

    val exercisesFilter: ExercisesFilters?
        get() = this.filterMutableLiveData.value

    init {
        val filter = ExercisesFilters()
        this.filterMutableLiveData.value = filter

        val source = Transformations.switchMap(
            this.filterMutableLiveData
        ) {repository.filterSelect(this.filterMutableLiveData.value!!)}

        this.data.addSource(source) {this.data.setValue(it)}
    }

    fun selectData(): LiveData<List<Exercise>> {
        return data
    }

    fun select(exerciseId: Int): LiveData<Exercise> {
        return this.repository.select(exerciseId)
    }

    fun insert(exercise: Exercise) {
        this.repository.insert(exercise)
    }

    fun setFilterMutableLiveData(exercisesFilter: ExercisesFilters) {
        this.filterMutableLiveData.value = exercisesFilter
    }
}
